lib/Backend/LowerMDShared.cpp

```diff
@@ -5812,6 +5812,165 @@ LowererMD::GenerateCtz(IR::Instr * instr)
+    IR::Opnd * twoTo31Float = MaterializeConstFromBits(_mm_extract_epi32(X86_TWO_31_F4.m128i_value, 0), TyFloat32, instr);
+
+    if (instr->GetDst()->IsUInt32())
+    {
+        limitReg = IR::RegOpnd::New(TyFloat64, m_func);
+        EmitFloat32ToFloat64(limitReg, twoTo31Float, instr);
+        IR::RegOpnd* maxIntDoubleReg = MaterializeDoubleConstFromInt(max, instr);
+        instr->InsertBefore(IR::Instr::New(Js::OpCode::ADDPD, limitReg, limitReg, maxIntDoubleReg, m_func));
+    }
+    else
+    {
+        limitReg = MaterializeDoubleConstFromInt(max, instr);
+    }
+
+    GenerateCheck(Js::OpCode::COMISD, limitReg, src64, Js::OpCode::JAE, conversion, instr); // >= max
+    instr->InsertBefore(throwLabel);
+    IR::Instr *throwInstr = IR::Instr::New(
+        Js::OpCode::RuntimeTypeError,
+        IR::RegOpnd::New(TyMachReg, m_func),
+        IR::IntConstOpnd::New(SCODE_CODE(VBSERR_Overflow), TyInt32, m_func),
+        m_func);
+    instr->InsertBefore(throwInstr);
+    this->m_lowerer->LowerUnaryHelperMem(throwInstr, IR::HelperOp_RuntimeTypeError);
+    //no jump here we aren't coming back
+
+    instr->InsertBefore(conversion);
+}
+
+IR::Opnd*
+LowererMD::TruncLower31BitsFlt(IR::Instr* instr, IR::Opnd* src1, IR::Opnd* dst, IR::Opnd* twoTo31Float)
+{
+    IR::RegOpnd * tmpReg = IR::RegOpnd::New(TyFloat32, m_func);
+    IR::Instr* cmp = IR::Instr::New(Js::OpCode::CMPLTPS, tmpReg, twoTo31Float, src1, m_func);
+
+    instr->InsertBefore(cmp);
+    Legalize(cmp);
+
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::ANDPS, tmpReg, tmpReg, twoTo31Float, m_func));
+    IR::Opnd * result = IR::RegOpnd::New(TyFloat32, m_func);
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::MOVAPS, result, src1, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::SUBPS, result, result, tmpReg, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::CVTTSS2SI, dst, result, m_func));
+
+    return tmpReg;
+}
+
+IR::Opnd*
+LowererMD::TruncLower31BitsDb(IR::Instr* instr, IR::Opnd* src1, IR::Opnd* dst, IR::Opnd* twoTo31Float)
+{
+    IR::RegOpnd * tmpReg = IR::RegOpnd::New(TyFloat64, m_func);
+    IR::Opnd * result = IR::RegOpnd::New(TyFloat64, m_func);
+    EmitFloat32ToFloat64(result, twoTo31Float, instr);
+    IR::Instr* cmp = IR::Instr::New(Js::OpCode::CMPLTPD, tmpReg, result, src1, m_func);
+    instr->InsertBefore(cmp);
+    Legalize(cmp);
+
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::ANDPD, tmpReg, tmpReg, result, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::MOVAPS, result, src1, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::SUBPD, result, result, tmpReg, m_func));
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::CVTTSD2SI, dst, result, m_func));
+
+    instr->InsertBefore(IR::Instr::New(Js::OpCode::CVTSD2SS, twoTo31Float, tmpReg, m_func));
+
+    return twoTo31Float;
+}
+
+void
+LowererMD::GenerateTruncWithCheck(IR::Instr * instr)
```

Comment:
Doesn't your code have more branches? Mine only has 1 (not taken) branch in the main case (helper block will be moved past ret in layout phase so main path will just have cvtss2si dst, src1; cmp src1, 80000000H; je helper; )

---
In reply to: [84940996](https://github.com/Microsoft/ChakraCore/pull/1812#discussion_r84940996) [](ancestors = 84940996)